cmd/skopeo/list_tags.go

```diff
@@ -113,23 +114,51 @@ func (opts *tagsOptions) run(args []string, stdout io.Writer) (retErr error) {
 		return fmt.Errorf("Invalid %q: does not specify a transport", args[0])
 	}
 
-	if transport.Name() != docker.Transport.Name() {
-		return fmt.Errorf("Unsupported transport '%v' for tag listing. Only '%v' currently supported", transport.Name(), docker.Transport.Name())
-	}
-
-	// Do transport-specific parsing and validation to get an image reference
-	imgRef, err := parseDockerRepositoryReference(args[0])
-	if err != nil {
-		return err
-	}
-
 	var repositoryName string
 	var tagListing []string
-	if err = retry.RetryIfNecessary(ctx, func() error {
-		repositoryName, tagListing, err = listDockerTags(ctx, sys, imgRef)
-		return err
-	}, opts.retryOpts); err != nil {
-		return err
+
+	switch transport.Name() {
+	case docker.Transport.Name():
+		// Do transport-specific parsing and validation to get an image reference
+		imgRef, err := parseDockerRepositoryReference(args[0])
+		if err != nil {
+			return err
+		}
+		if err = retry.RetryIfNecessary(ctx, func() error {
+			repositoryName, tagListing, err = listDockerTags(ctx, sys, imgRef)
+			return err
+		}, opts.retryOpts); err != nil {
+			return err
+		}
+	case archive.Transport.Name():
+		ref, err := archive.ParseReference(strings.TrimPrefix(args[0], archive.Transport.Name()+":"))
+		if err != nil {
+			return err
+		}
+		archivePath := ref.StringWithinTransport()
+		tarReader, err := archive.NewReader(sys, archivePath)
+		if err != nil {
+			return err
+		}
+		defer tarReader.Close()
+
+		imageRefs, err := tarReader.List()
+		if err != nil {
+			return err
+		}
+
+		for _, items := range imageRefs {
+			for _, ref := range items {
+				repoTags, err := tarReader.ManifestTagsForReference(ref)
+				if err != nil {
+					return err
+				}
+				tagListing = append(tagListing, repoTags...)
+			}
+		}
+	default:
+		return fmt.Errorf("Unsupported transport '%v' for tag listing. Only supported: %q,%q ",
+			transport.Name(), docker.Transport.Name(), archive.Transport.Name())
```

Comment:
Non-blocking?

Define a
```go
transportHandlers map[transportName string]
    func(userInput string) (repositoryName string, tagListing []string, error)
```
, and split handling of the different transports into separate functions. That way it scales better than one large function.

Then, in this (“no map entry found”) case, we can programmatically iterate the map keys, and construct a sorted list of the supported transports in a way that never gets out of sync.

OTOH at 2 cases this is not urgent, I’d insist on this at about 3 cases.

Reply:
I tried to change to this https://paste.debian.net/1232877/ ,  But it doesn't look good

Reply:
Oh, the functions in `transportHandlers` would of course be separate named functions, not in-line lambdas.

```go
var transportHandlers = map[string]func(…) (…) {
    docker.Transport.Name(): listDockerTags,
    archive.Transport.Name(): listDockerArchiveTags,
}
```